File: crates/taflow-core/src/indicators/math_cbrt.rs
//! Persistent pointwise `cbrt` transform.

use crate::error::TaResult;
use crate::stream::StreamingIndicator;

/// Apply `cbrt` to each value without warm-up.
#[derive(Debug, Clone, Default)]
pub struct MathCbrt {
    value: Option<f64>,
}

impl MathCbrt {
    /// Create a fresh pointwise transform state.
    pub fn new() -> TaResult<Self> {
        Ok(Self::default())
    }

    /// Transform one chronological value.
    pub fn append(&mut self, input: f64) -> Option<f64> {
        self.value = Some(input.cbrt());
        self.value
    }

    /// Transform a slice into `output` while preserving scalar replay order.
    pub fn extend_slice_into(&mut self, input: &[f64], output: &mut Vec<f64>) {
        output.reserve(input.len());
        output.extend(input.iter().map(|&input| {
            self.append(input)
                .expect("pointwise transforms have no warm-up")
        }));
    }

    /// Return the latest result, or `None` before the first value.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Restore fresh-state behavior without reallocating.
    pub fn reset(&mut self) {
        self.value = None;
    }
}

impl StreamingIndicator for MathCbrt {
    type Output = f64;

    fn append(&mut self, input: f64) -> Option<Self::Output> {
        Self::append(self, input)
    }

    fn value(&self) -> Option<Self::Output> {
        Self::value(self)
    }

    fn reset(&mut self) {
        Self::reset(self);
    }
}

File: python/taflow/indicators/math_cbrt.py
"""Persistent pointwise cbrt transform."""

from typing import Any

from .._math_state import MathUnaryState
from .._native import MathCbrt as _NativeMathCbrt


class MathCbrt(MathUnaryState):
    """Apply pointwise cbrt in persistent Rust state.

    Constructing the class creates a fresh empty state. Supply chronological
    values through ``extend`` or ``append``.

    The output is a same-length ``float64`` array with no rolling warm-up.
    Domain behavior follows IEEE 754. The independent correctness oracle is
    ``np.cbrt``.
    """

    _native_cls = _NativeMathCbrt

    def append(self, _input: float) -> "MathCbrt":
        """Append one value and return this indicator."""
        super().append(_input)
        return self

    def extend(self, _input: Any) -> "MathCbrt":
        """Append chronological values and return this indicator."""
        super().extend(_input)
        return self

    def reset(self) -> "MathCbrt":
        """Restore fresh native state and return this indicator."""
        super().reset()
        return self
